File: second-round/283.移动零.py
#
# @lc app=leetcode.cn id=283 lang=python3
#
# [283] 移动零
#
# https://leetcode-cn.com/problems/move-zeroes/description/
#
# algorithms
# Easy (64.03%)
# Likes:    1395
# Dislikes: 0
# Total Accepted:    595.7K
# Total Submissions: 930.4K
# Testcase Example:  '[0,1,0,3,12]'
#
# 给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。
# 
# 示例:
# 
# 输入: [0,1,0,3,12]
# 输出: [1,3,12,0,0]
# 
# 说明:
# 
# 
# 必须在原数组上操作，不能拷贝额外的数组。
# 尽量减少操作次数。
# 
# 
#

# @lc code=start
class Solution:
    def moveZeroes(self, nums: list[int]) -> None:
        i, p = 0, 0

        while i < len(nums): # F IXME
            while i + 1 < len(nums) and nums[i] == 0:
                i += 1
            nums[p] = nums[i]
            i += 1
            p += 1

        for i in range(p, len(nums)):
            nums[i] = 0
# @lc code=end

q = Solution()
s = [0,1,0,3,12]
q.moveZeroes(s)
print(s)
s = [0,0,0,0,0]
q.moveZeroes(s)
print(s)
s = []
q.moveZeroes(s)
print(s)
s = [1,2,0,0]
q.moveZeroes(s)
print(s)

File: 6/283.移动零.ts
/*
 * @lc app=leetcode.cn id=283 lang=typescript
 *
 * [283] 移动零
 *
 * https://leetcode.cn/problems/move-zeroes/description/
 *
 * algorithms
 * Easy (64.03%)
 * Likes:    1633
 * Dislikes: 0
 * Total Accepted:    788.6K
 * Total Submissions: 1.2M
 * Testcase Example:  '[0,1,0,3,12]'
 *
 * 给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。
 *
 * 请注意 ，必须在不复制数组的情况下原地对数组进行操作。
 *
 *
 *
 * 示例 1:
 *
 *
 * 输入: nums = [0,1,0,3,12]
 * 输出: [1,3,12,0,0]
 *
 *
 * 示例 2:
 *
 *
 * 输入: nums = [0]
 * 输出: [0]
 *
 *
 *
 * 提示:
 *
 *
 *
 * 1 <= nums.length <= 10^4
 * -2^31 <= nums[i] <= 2^31 - 1
 *
 *
 *
 *
 * 进阶：你能尽量减少完成的操作次数吗？
 *
 */

export
// @lc code=start
function moveZeroes(nums: number[]): void {
  let p = 0

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === 0) {
      continue
    }
    nums[p++] = nums[i]
  }

  while (p < nums.length) {
    nums[p++] = 0
  }
}
// @lc code=end
